File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        brackets = {'(': ')', '{': '}', '[': ']'}
        stack = []
        
        for c in s:
            if c in brackets:
                stack.append(c)
            else:
                if stack and brackets[stack[-1]] == c:
                    stack.pop()
                else:
                    return False
        
        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        map<char, char> brackets{{'(', ')'}, {'{', '}'}, {'[', ']'}};
        stack<char> stack;

        for (char c: s) {
            if (!brackets.contains(c)) {
                if (stack.empty() || c != brackets[stack.top()]) {
                    return false;
                } else {
                    stack.pop();
                }
            } else {
                stack.push(c);
            }
        }

        return stack.empty();
    }
};